### PR TITLE
[Bugfix] Invalid generation of types with >1 generic parameters

### LINKED ISSUE #550: Invalid generation of types with >1 generic
### Description
(a). If a type has >1 generic, the >1 types are considered phantoms.
(b). More than considering them phantoms, ALL the fields using >1 generic in index are coerced back to the 0th one (i.e. with the below example, `a, b, c, d: T`. This is caused by (d).
(c). This result in non-derived Encode/Decode for the given type.

(d). Worth to mention, the `TrackedSymbol` ids are all the same and it looks like they are comming from `TypeInfo` derive macro. Not sure this is a normal behavior either.

I'm currently trying to fix the issue.

### Reproducible snippet
```rust
#[allow(unused)]
#[derive(TypeInfo)]
struct Foo<T, U, V, W> {
    a: T,
    b: U,
    c: V,
    d: W
}
```

### Test case result
```
Diff < left / right > :
<pub mod tests { use super :: root ; # [derive (:: subxt :: codec :: Decode , :: subxt :: codec :: Encode , Debug)] pub struct Bar { pub p : root :: subxt_codegen :: types :: tests :: Foo < :: core :: primitive :: u32 , :: core :: primitive :: u32 , :: core :: primitive :: u64 , :: core :: primitive :: u128 > , pub q : root :: subxt_codegen :: types :: tests :: Foo < :: core :: primitive :: u8 , :: core :: primitive :: u8 , :: core :: primitive :: u8 , :: core :: primitive :: u8 > , } # [derive (:: subxt :: codec :: Decode , :: subxt :: codec :: Encode , Debug)] pub struct Foo < _0 , _1 , _2 , _3 > { pub a : _0 , pub b : _0 , pub c : _0 , pub d : _0 , # [codec (skip)] pub __subxt_unused_type_params : :: core :: marker :: PhantomData < (_1 , _2 , _3) > } }
>pub mod tests { use super :: root ; # [derive (:: subxt :: codec :: Decode , :: subxt :: codec :: Encode , Debug)] pub struct Bar { pub p : root :: subxt_codegen :: types :: tests :: Foo < :: core :: primitive :: u32 , :: core :: primitive :: u32 , :: core :: primitive :: u64 , :: core :: primitive :: u128 > , pub q : root :: subxt_codegen :: types :: tests :: Foo < :: core :: primitive :: u8 , :: core :: primitive :: u8 , :: core :: primitive :: u8 , :: core :: primitive :: u8 > , } # [derive (:: subxt :: codec :: Decode , :: subxt :: codec :: Encode , Debug)] pub struct Foo < _0 , _1 , _2 , _3 > { pub a : _0 , pub b : _1 , pub c : _2 , pub d : _3 , } }
```

### Logs from the test
```
[2022-05-26T20:47:33Z INFO  subxt_codegen::types::type_def] Type "Foo" fields [Field { name: Some("a"), ty: UntrackedSymbol { id: 4, marker: PhantomData }, type_name: Some("T"), docs: [] }, Field { name: Some("b"), ty: UntrackedSymbol { id: 4, marker: PhantomData }, type_name: Some("U"), docs: [] }, Field { name: Some("c"), ty: UntrackedSymbol { id: 4, marker: PhantomData }, type_name: Some("V"), docs: [] }, Field { name: Some("d"), ty: UntrackedSymbol { id: 4, marker: PhantomData }, type_name: Some("W"), docs: [] }] generics [TypeParameter { concrete_type_id: 4, original_name: "T", name: Ident(_0) }, TypeParameter { concrete_type_id: 4, original_name: "U", name: Ident(_1) }, TypeParameter { concrete_type_id: 4, original_name: "V", name: Ident(_2) }, TypeParameter { concrete_type_id: 4, original_name: "W", name: Ident(_3) }]
```

### PR DESCRIPTION
**Pull request by tadeohepperle**
fixes #550

**Investigated reason for the bug:** In subxt, when calling `resolve_field_type_path()` to get the type of a generic, we were not taking into account multiple generics with different names (`originial_name` e.g. `T`, `U`, `V`, ...). The function just looked for the first type where the type (e.g. `u8`) matches (see: `parent_type_params.iter().find(...)`). 

**Fix:** add an optional parameter `original_name: Option<&str>` to the type resolving function to also check the name of the generic argument.

Added a test to confirm the behavior is now as desired. The test matches the problem statement in #550.

### CODE REVIEW

**Review by jsdw — APPROVED** [review]
Nice job! Also love the test :)

File: testing/integration-tests/src/codegen/codegen_tests.rs [hunk 1]
```diff
@@ -144,3 +145,75 @@ fn generic_types_overwrite_each_other() {
     // We do _not_ expect this to exist, since a generic is present on the type:
     assert!(!interface.contains("DuplicateType2"));
 }
+
+#[test]
+fn more_than_1_generic_parameters_work() {
```
Comment (lexnv) [inline testing/integration-tests/src/codegen/codegen_tests.rs:150]: Nice!

**Review by lexnv — APPROVED** [review]
Looks good! Amazing PR!

File: testing/integration-tests/src/codegen/codegen_tests.rs [hunk 2]
```diff
@@ -6,8 +6,9 @@ use frame_metadata::{
     v15::{ExtrinsicMetadata, RuntimeMetadataV15},
     RuntimeMetadataPrefixed,
 };
-use scale_info::{meta_type, IntoPortable, TypeInfo};
+use scale_info::{meta_type, IntoPortable, PortableRegistry, Registry, TypeInfo};
 use subxt_codegen::{CratePath, DerivesRegistry, RuntimeGenerator, TypeSubstitutes};
+use syn::__private::quote;
```
[the remark this answers is not in the document]
Comment (tadeohepperle) [reply testing/integration-tests/src/codegen/codegen_tests.rs:11]: I could do that, good idea.

**Pull request merged**